src/main/scala/io/qbeast/spark/delta/SamplingListFilesStrategy.scala

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2021 Qbeast Analytics, S.L.
+ */
+package io.qbeast.spark.delta
+
+import io.qbeast.spark.index.query.QueryExecutor
+import io.qbeast.spark.index.query.QuerySpecBuilder
+import io.qbeast.spark.utils.TagUtils
+import org.apache.hadoop.fs.FileStatus
+import org.apache.hadoop.fs.Path
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow
+import org.apache.spark.sql.delta.actions.AddFile
+import org.apache.spark.sql.delta.files.TahoeLogFileIndex
+import org.apache.spark.sql.delta.Snapshot
+import org.apache.spark.sql.execution.datasources.FileStatusWithMetadata
+import org.apache.spark.sql.execution.datasources.PartitionDirectory
+import org.apache.spark.sql.execution.SQLExecution
+
+/**
+ * Implementation of ListFilesStrategy to be used when the query contains sampling clauses.
+ */
+private[delta] object SamplingListFilesStrategy
+    extends ListFilesStrategy
+    with Logging
+    with Serializable {
+
+  override def listFiles(
+      target: TahoeLogFileIndex,
+      partitionFilters: Seq[Expression],
+      dataFilters: Seq[Expression]): Seq[PartitionDirectory] = {
+    val snapshot = target.getSnapshot
+    val path = target.path
+    val files = listStagingAreaFiles(target, partitionFilters, dataFilters) ++
+      listIndexFiles(snapshot, path, partitionFilters, dataFilters)
+    logFilteredFiles(target, snapshot, files)
+    Seq(PartitionDirectory(new GenericInternalRow(Array.empty[Any]), files))
+  }
+
+  private def listStagingAreaFiles(
+      target: TahoeLogFileIndex,
+      partitionFilters: Seq[Expression],
+      dataFilters: Seq[Expression]): Seq[FileStatusWithMetadata] = {
+    target
+      .matchingFiles(partitionFilters, dataFilters)
+      .filter(isStagingAreaFile)
+      .map(file =>
+        new FileStatus(
+          file.size,
+          false,
+          0,
+          1,
+          file.modificationTime,
+          getAbsolutePath(target, file)))
+      .map(file => new FileStatusWithMetadata(file, Map.empty))
+  }
+
+  private def isStagingAreaFile(file: AddFile): Boolean = file.getTag(TagUtils.revision) match {
```

Comment:
I think there is some trait or object called `StagingUtils` where this method can be more useful.

Reply:
There are two traits StagingUtils in the qbeast-core, and DeltaStagingUtils in qbeast-spark. The first one knows nothing about AddFile, and the second one has method I need ´isStagingFile´ however implementing DeltaStagingUtils in SamplingListFileStrategy is hard, because it requires to implement ´snapshot´ method returning the Delta Snapshot, but the design of ListFileStrategy is that there is no state, the snapshot is computed on the stack. Probably we need to redesign DeltaStagingUtils first.